sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java

```diff
@@ -758,4 +789,43 @@ private static void logAddressResolutionEnd(RxDocumentServiceRequest request, St
             BridgeInternal.recordAddressResolutionEnd(request.requestContext.cosmosDiagnostics, identifier, errorMessage);
         }
     }
+
+    private static class ForcedRefreshMetadata {
+        private final ConcurrentHashMap<PartitionKeyRangeIdentity, Instant> lastPartitionAddressOnlyRefresh;
+        private Instant lastCollectionRoutingMapRefresh;
+
+        public ForcedRefreshMetadata() {
+            lastPartitionAddressOnlyRefresh = new ConcurrentHashMap<>();
+            lastCollectionRoutingMapRefresh = Instant.now();
+        }
+
+        public void signalCollectionRoutingMapRefresh(PartitionKeyRangeIdentity pk) {
+            Instant nowSnapshot = Instant.now();
+            lastPartitionAddressOnlyRefresh.put(pk, nowSnapshot);
+            lastCollectionRoutingMapRefresh = nowSnapshot;
+        }
+
+        public void signalPartitionAddressOnlyRefresh(PartitionKeyRangeIdentity pk) {
+            lastPartitionAddressOnlyRefresh.put(pk, Instant.now());
+        }
+
+        public boolean shouldIncludeCollectionRoutingMapRefresh(PartitionKeyRangeIdentity pk) {
```

Comment:
As per the discussion we decided to refresh collection routing map for same request on second retry.
Now this class hold the state regardless of requests, therefore one single 410 retry for a request make collectionRoutingMap refresh for another request even on single 410 if it comes after 30 sec., is this expected ?

Overall the state to check collectionRoutingMap  refresh depend on a request retries ?

Reply:
I will only allow triggering the collection routing map refresh for a partition where we recently tried just refreshing the addresses. And only once every 30 seconds per container. I think that is safe - and reflects our original plan by also allowing to avoid a major refactoring of the code (because we simply don't have a place to do this based on request level retries without large refactoring)

Reply:
One more question on top of this.

when we setup the HttpHeaders.FORCE_COLLECTION_ROUTING_MAP_REFRESH, is that a guarantee that gateway will return the updated address? 

And what if the request to refresh the address failed(maybe due to network issue)? will this block all other tries in the next 30s?

Reply:
Expectation is that in this case Gateway would return error that the PKRangeId doesn't exist anymore. Not guaranteed to happen sync - but it will trigger updating the cache to get to this behavior.
And yes - it will block us setting the CollectionRefresh (if not explicitly asked for otherwise) - basically the proactive collection refresh - that is the trade-off of the throttling on our side. But given that we never recovered before - recovering after 30 seconds is at least a step in the right direction :-)

The real fix is to not rely on throttling on the client but have cache refresh and throttling in the gateway.